### PR TITLE
[NT-1898] Retry Failed Comment And Refactor Comment Posting

### PR DESCRIPTION
**Pull request by justinswart**
# 📲 What

Adds retrying of failed comments.

⚠️ This stuff is hard to do with FRP and adds significant complexity to this view model ⚠️ 

# 🤔 Why

We've built the new comments posting using an optimistic UI approach. When posting the UI immediately reflects the new comment and _optimistically_ expects the call to the back-end to succeed.

# 🛠 How

When posting a temporary `Comment` is immediately inserted into the datasource. If posting fails, that `Comment` is replaced with an identical value with its `status` updated to `failed`. Tapping the comment updates its status to `retrying` and upon success flips it temporarily to `retrySuccess` and then `success` which is the regular status. If retrying fails the status is set back to `failed` and the process repeats.

# 👀 See

In this video I am tapping the 3 errored comments in succession:

https://user-images.githubusercontent.com/3735375/120873212-0b289380-c556-11eb-989e-ba7dca7d9e8b.mp4

# ✅ Acceptance criteria

- [x] Using Charles proxy, set a breakpoint and try to post a new comment, when the breakpoint is hit abort the request so that it fails. Remove the breakpoint and tap on the failed comment - it should post again and, if successful, be replaced in the list.
- [x] Try the previous test with multiple failures.
- [x] Test the happy path and page in more comments, pull to refresh, etc, there should be no regression on the previous behaviour.
- [ ] @moderateepheezy and @msadoon to please confirm that I didn't lose any of their work during resolving merge conflicts.

# ⏰ TODO

- [ ] May need another test for the multiple deletion, just need to think through it a bit more.

### CODE REVIEW

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by msadoon — APPROVED** [review]
This is incredible! Yea it took me a while to figure out the order of the signals, but the tests are really thorough. I still haven't tested the on simulator, but will do that today and post here if there's any issues.

File: Library/ViewModels/CommentsViewModel.swift [hunk 1]
```diff
@@ -105,99 +99,144 @@ public final class CommentsViewModel: CommentsViewModelType,
...
-      .combineLatest(comments, optimisticPostComment)
-      .map { loadedComments, optimisticComment -> [Comment] in
-        [optimisticComment] + loadedComments
-      }.map { $0 as Array }
+    let commentsWithFailableOrComment = currentComments
+      .takePairWhen(self.failableOrComment.signal.skipNil())
+      .map(unpack)
+      .map(commentsReplacingCommentById)
+
+    self.resetCommentComposerAndScrollToTop = commentsWithFailableOrComment
```
[the remark this answers is not in the document]
Comment (singhhari) [reply Library/ViewModels/CommentsViewModel.swift:156]: yeah, the naming is admittedly a bit hard to follow along with.

File: Library/AppEnvironment.swift [hunk 2]
```diff
@@ -153,7 +153,8 @@ public struct AppEnvironment: AppEnvironmentType {
     reachability: SignalProducer<Reachability, Never> = AppEnvironment.current.reachability,
     scheduler: DateScheduler = AppEnvironment.current.scheduler,
     ubiquitousStore: KeyValueStoreType = AppEnvironment.current.ubiquitousStore,
-    userDefaults: KeyValueStoreType = AppEnvironment.current.userDefaults
+    userDefaults: KeyValueStoreType = AppEnvironment.current.userDefaults,
+    uuidType: UUIDType.Type = AppEnvironment.current.uuidType
```
[the remark this answers is not in the document]
Comment (singhhari) [reply Library/AppEnvironment.swift:157]: Nvm, I see why in the view model.

File: Library/UUIDType.swift [hunk 3]
```diff
@@ -2,6 +2,7 @@ import Foundation
 
 public protocol UUIDType {
   var uuidString: String { get }
+  init()
```
[the remark this answers is not in the document]
Comment (justinswart) [reply Library/UUIDType.swift:5]: so this protocol represents the _type_ and not the instance of the type, so that allows to instantiate a new `UUID()` or `MockUUID()` in our view models which helps to make assertions based on the ID. If we called `UUID()` directly we'd of course get a random `UUID` each time.

File: Library/ViewModels/CommentsViewModel.swift [hunk 4]
```diff
@@ -220,17 +259,93 @@ public final class CommentsViewModel: CommentsViewModelType,
...
+  .map { replacementComment in (replacementComment, comment.id) }
+}
+
+private func postCommentProducer(
+  project: Project,
+  user: User,
+  body: String
+) -> SignalProducer<(Comment, String), Never> {
+  let failableComment = Comment.failableComment(
+    withId: AppEnvironment.current.uuidType.init().uuidString,
```
[the remark this answers is not in the document]
Comment (justinswart) [reply Library/ViewModels/CommentsViewModel.swift:311]: Yeah `.self` is to set this value to `UUID.self` which is `UUID.Type`.

File: Library/ViewModels/CommentsViewModel.swift [hunk 5]
```diff
@@ -105,99 +99,144 @@ public final class CommentsViewModel: CommentsViewModelType,
...
-
-    // TODO: Handle error and success states appropriately for the datasource item
-    self.postCommentSuccessful = postCommentEvent.values().map { $0 }
-    self.errorMessage = postCommentEvent.errors().map { $0.errorMessages.first }.skipNil()
-
     let isCloseToBottom = self.willDisplayRowProperty.signal.skipNil()
-      .map { row, total in row >= total - 3 }
+      .map { row, total -> Bool in
+        // Prevent paging when only the empty state cell is shown.
+        guard total > 1 else { return false }
```
[the remark this answers is not in the document]
Comment (justinswart) [reply Library/ViewModels/CommentsViewModel.swift:105]: Hmm, yeah maybe - I might just add a `TODO` here for this to be improved when we do the pagination error handling because I think that'll touch this part.

File: Library/ViewModels/CommentsViewModel.swift [hunk 6]
```diff
@@ -105,99 +99,144 @@ public final class CommentsViewModel: CommentsViewModelType,
...
 
+    let pullToRefresh = self.refreshProperty.signal
+
     let requestFirstPageWith = Signal.merge(
       initialProject,
-      initialProject.takeWhen(self.refreshProperty.signal),
-      initialProject.takeWhen(self.postCommentSuccessful)
+      initialProject.takeWhen(pullToRefresh)
+        // Thread hop so that we can clear our comments buffer before newly paginated results.
+        .ksr_debounce(.nanoseconds(0), on: AppEnvironment.current.scheduler)
```
[the remark this answers is not in the document]
Comment (justinswart) [reply Library/ViewModels/CommentsViewModel.swift:118]: It's kind've like calling this using `DispatchQueue.main.async { ... }`. It runs on the main thread but on the next run loop. Helped fix some deadlocks I was having.

File: Library/ViewModels/CommentsViewModel.swift [hunk 7]
```diff
@@ -105,99 +99,144 @@ public final class CommentsViewModel: CommentsViewModelType,
...
+    let commentsAndProject = initialProject
+      .takePairWhen(paginatedComments)
+      .map { ($1, $0) }
 
-    self.loadCommentsAndProjectIntoDataSource = commentsAndProject
-    self.isCommentsLoading = isLoading
-    self.isCellSeparatorHidden = cellSeparatorHidden
+    self.currentComments <~ commentsAndProject.map(first)
+      // Thread hop so that we don't circularly buffer.
+      .ksr_debounce(.nanoseconds(0), on: AppEnvironment.current.scheduler)
```
[the remark this answers is not in the document]
Comment (justinswart) [reply Library/ViewModels/CommentsViewModel.swift:186]: Same answer as https://github.com/kickstarter/ios-oss/pull/1501#discussion_r646883865

File: Library/ViewModels/CommentsViewModel.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (justinswart) [reply Library/ViewModels/CommentsViewModel.swift:156]: Hmm yeah but it is emitting a failable (optimistic) _or_ a comment (successfully posted).

**Review by moderateepheezy — APPROVED** [review]
It works as expected. I just need to through the implementation again to understand it better.

File: Library/ViewModels/CommentsViewModel.swift [hunk 8]
```diff
@@ -105,99 +99,144 @@ public final class CommentsViewModel: CommentsViewModelType,
...
+    let commentsWithFailableOrComment = currentComments
+      .takePairWhen(self.failableOrComment.signal.skipNil())
+      .map(unpack)
+      .map(commentsReplacingCommentById)
+
+    self.resetCommentComposerAndScrollToTop = commentsWithFailableOrComment
+      // We only want to scroll to top for failable comments as they represent the initial
+      // comment that's inserted. We check here to see that the first comment's ID is a UUID
+      // to determine this. There may be better ways.
+      .map { UUID(uuidString: $0.first?.id ?? "") }
```
[the remark this answers is not in the document]
Comment (justinswart) [reply Library/ViewModels/CommentsViewModel.swift:160]: Yeah but this looks at the most recent to see if it is a failable comment to know if it should scroll to the top. In the scenario you describe it should not scroll to the top.

File: KsApi/models/Comment.swift [hunk 9]
```diff
@@ -34,17 +31,21 @@ public struct Comment {
     case you
   }
 
-  public enum Status {
+  public enum Status: String, Decodable {
     case failed
-    case removed
+    case retrying
```
[the remark this answers is not in the document]
Comment (justinswart) [reply KsApi/models/Comment.swift:36]: `retrying` is replaced by `retrySuccess` after the post is successfully retried, it's then replaced one second later by `success`. This allows the cells to transition through their various states for showing progress in the UI.

File: KsApi/models/Comment.swift [hunk 10]
```diff
@@ -59,16 +60,19 @@ extension Comment {
...
+      createdAt: date.timeIntervalSince1970,
+      id: id,
       isDeleted: false,
-      replyCount: 0
+      replyCount: 0,
+      status: .success
     )
   }
+
+  public func updatingStatus(to status: Comment.Status) -> Comment {
```
[the remark this answers is not in the document]
Comment (justinswart) [reply KsApi/models/Comment.swift:71]: I expect to use it again for replies.

File: Library/ViewModels/CommentsViewModel.swift [hunk 11]
```diff
@@ -220,17 +259,93 @@ public final class CommentsViewModel: CommentsViewModelType,
...
+    project: project,
+    user: user,
+    body: body
+  )
+
+  // Post the new comment.
+  return AppEnvironment.current.apiService.postComment(
+    input: .init(
+      body: body,
+      commentableId: project.graphID
```
[the remark this answers is not in the document]
Comment (justinswart) [reply Library/ViewModels/CommentsViewModel.swift:323]: > What is `GraphIDBridging`?

Just a protocol that ensures that this type can construct a relay ID for matching in our schema. Relay IDs have the structure of `Model-<id>` so in this case like `Comment-12345`.

[the remark this answers is not in the document]
Comment (justinswart) [reply Library/ViewModels/CommentsViewModel.swift:323]: > In general, I was just wondering how the `id` of the `Comment` is created before its' `failed` status is confirmed by the network request? Because my assumption was a `Comment` has no `id` until it is `successful`. Just my own knowledge gap there.

This refers to the ID of the project to associate the comment to it. The project is the `Commentable` here.

**Comment by msadoon** [comment]
**on device testing notes**

Can confirm empty state appears when there are no comments and upon refreshing still contains the "failed" comment, so it can be re-tried.
One thing I that is sort of an edge case:
Pull to refresh removes a failed comment when it's not the only comment in the list.| Pull to refresh doesn't remove a failed comment when it's the only comment in the list.|
|---|---|
|||

Other than that, looks good to me.

**Comment by justinswart** [comment]
> Pull to refresh removes a failed comment when it's not the only comment in the list.

Interesting! Good edge-case to test. I think in the interest of unblocking other work I'm going to address this in a separate ticket 👍

**Review by singhhari — APPROVED** [review]
